### PR TITLE
Add support for memcached PECL library.

### PR DESCRIPTION
**Pull request by crebelein**
As previously discussed in pull request #320, I need support for the new memcached extension (https://pecl.php.net/package/memcached). This PR is a much more elegant solution to the problem that eliminates code duplication yet still allows for simple future enhancements when applicable.

### CODE REVIEW

File: adodb-memcache.lib.inc.php [hunk 1]
```diff
@@ -28,17 +28,21 @@
...
+  Notes; The memcache class is shared by all connections, is created during the first call to Connect/PConnect.
+         We'll look for both the memcache library (https://pecl.php.net/package/memcache) and the  memcached
+         library (https://pecl.php.net/package/memcached). If both exist, the memcache library will be used.
 
   Class instance is stored in $ADODB_CACHE
 */
 
 	class ADODB_Cache_MemCache {
 		var $createdir = false; // create caching directory structure?
+		var $library = false; // check for memcache/memcached prior to connect, then populate this variable
```
Comment (dregad) [inline adodb-memcache.lib.inc.php]: This comment is not very clear; when I read it, I have the impression that I need to manually set it. If that's the case, I think it needs to be better documented. But the code in connect() method initializes it...

File: adodb-memcache.lib.inc.php [hunk 2]
```diff
@@ -93,8 +102,23 @@ function writecache($filename, $contents, $debug, $secs2cache)
...
+					if (!$this->_memcache->set($filename, $contents, $this->compress ? MEMCACHE_COMPRESSED : 0, $secs2cache)) {
+						$failed=true;
+					}
+					break;
+				case 'Memcached':
+					if (!$this->_memcache->set($filename, $contents, $secs2cache)) {
+						$failed=true;
+					}
+					break;
+				default: $failed=true; break;
```
Comment (dregad) [inline adodb-memcache.lib.inc.php]:
```
default:
    $failed=true;
    break;
```

**Comment by dregad** [comment]
Looks good to me, with a couple minor comments. 

@mnewnham please check and follow up as per your ealier note https://github.com/ADOdb/ADOdb/pull/320#issuecomment-289936916.

**Comment by crebelein** [comment]
@dregad Very good point concerning the comment pertaining to $library. I've added some clarity to this. I've also made the formatting change in the switch block that you pointed out.

**Comment by dregad** [comment]
Thanks for the corrections. I'll merge this when I hear from @mnewnham. Alternatively, if you're willing to review and provide updates° for our documentation:

- http://adodb.org/dokuwiki/doku.php?id=v5:userguide:memcached
- http://adodb.org/dokuwiki/doku.php?id=v5:userguide:learn_caching:memcached
- also if you don't mind, documenting the _Undocumented series of options_ at bottom of table in http://adodb.org/dokuwiki/doku.php?id=v5:reference:connection:adonewconnection#additional_parameters_to_dsn_connection

° just post edited RAW Dokuwiki markup here

**Comment by crebelein** [comment]
Sure, I'm happy to augment the documentation. I'll start with this one and do the other 2 as soon as I am able.

http://adodb.org/dokuwiki/doku.php?id=v5:userguide:learn_caching:memcached

[learn_caching.txt](https://github.com/ADOdb/ADOdb/files/922594/learn_caching.txt)

**Comment by crebelein** [comment]
http://adodb.org/dokuwiki/doku.php?id=v5:userguide:memcached

[memcached.txt](https://github.com/ADOdb/ADOdb/files/922610/memcached.txt)

**Comment by mnewnham** [comment]
I'll try to check it tomorrow

**Comment by mnewnham** [comment]
There still appears to be no support for Windows with memcached, but the memcache dll is still actively maintained and there is a version for PHP 7.1 [here](https://github.com/nono303/PHP7-memcache-dll). I loaded it up and ran some test scripts and it still works fine. As far as I can tell from the  [memcached developer site](https://github.com/php-memcached-dev/php-memcached) there seems to be no interest in providing windows dll files, so I think that our documentation should be modified to reflect that the memcache library is not obsolete, but a parallel process for windows clients

**Comment by crebelein** [comment]
@mnewnham, thanks for pointing that out. I was curious about the state of Windows when I was editing the documentation. I'll redo the above docs and finish up the third, hopefully later tonight.

**Comment by crebelein** [comment]
Here is new documentation for the learn_caching page which take's @mnewnham's comments into consideration.

[learn_caching.txt](https://github.com/ADOdb/ADOdb/files/923146/learn_caching.txt)

**Comment by crebelein** [comment]
Here is new documentation for the memcached page which take @mnewnham's comments into consideration.

[memcached.txt](https://github.com/ADOdb/ADOdb/files/923148/memcached.txt)

**Comment by crebelein** [comment]
adoNewConnection page

[adonewconnection.txt](https://github.com/ADOdb/ADOdb/files/923158/adonewconnection.txt)

**Comment by mnewnham** [comment]
I've updated the documentation with the provided files, thank you

**Comment by dregad** [comment]
I made a few (mostly cosmetic) adjustments to the pages. Doing so, I realized that there is a lot of overlap between the 3 pages; 2 of them even have the same name... Wouldn't it be worth to refactor this and avoid duplicating information ?

**Comment by crebelein** [comment]
@dregad, I would agree. When I was reworking these pages I was almost cutting and pasting between the learn_caching and memcached pages. Merging them would make a lot of sense.

On another note, it looks like you cleaned up the bottom of the adoNewConnection page. I intended to point that out but it must have slipped my mind. Other adjustments look good as well.

**Comment by mnewnham** [comment]
I'd have to disagree on merging pages.  If the tutorial looks too much like a reference page then we need to improve the quality of the page from a 'tutorial' perspective, for example adding comparisons and longer explanations as well as improving the quality of the example to include retrieving a result set etc.  Page views of the tutorials are quite substantial and so we should make them as expansive as possible. This helps reduce the number of issues posted on github

**Comment by dregad** [comment]
That's a sound argument. From my perspective, the problem is not so much what the pages look like or the fact that tutorial is too terse, but the fact that the content is pretty much identical. 

In that case, maybe we could use the [DokuWiki include plugin](https://www.dokuwiki.org/plugin:include) instead of duplicating content ?

**Pull request merged**